APCAppCore/APCAppCore/Library/Scoring/APCScoring.m

```diff
@@ -281,37 +314,186 @@ - (instancetype)initWithHealthKitQuantityType:(HKQuantityType *)quantityType
                     groupBy:(APHTimelineGroups)groupBy
-      withCompletionHandler:(void (^)(void))completion
 {
-    NSInteger days = numberOfDays + 1;
+    
+    _groupBy = groupBy;
+    _numberOfDays = numberOfDays;
+    
+    __weak typeof(self) weakSelf = self;
+    
+    // Update the generated dataset to be inline with the timeline.
+    self.dataPoints = [self dataPointsArrayForDays:_numberOfDays groupBy:_groupBy];
     
     if (self.usesHealthKitData) {
         if ([self.quantityType isCompatibleWithUnit:self.hkUnit]) {
             
             [self updateStatsCollectionForQuantityType:self.quantityType
                                                   unit:self.hkUnit
-                                               forDays:days
+                                               forDays:numberOfDays
                                                groupBy:groupBy
-                                            completion:completion];
+                                            completion:^{
+                                                [weakSelf updateCharts];
+                                            }];
         } else {
             NSAssert([self.quantityType isCompatibleWithUnit:self.hkUnit], @"The quantity and the unit must be compatible");
         }
     } else {
-        // Update the timeline based on the number of days requested.
-        self.timeline = [self configureTimelineForDays:days groupBy:groupBy];
-        
-        // Update the generated dataset to be inline with the timeline.
-        [self generateEmptyDataset];
         
         [self queryTaskId:self.taskId
-                    forDays:days
+                  forDays:numberOfDays
                  valueKey:self.valueKey
                   dataKey:self.dataKey
                   sortKey:self.sortKey
-               latestOnly:YES
+               latestOnly:self.latestOnly
                   groupBy:groupBy
-               completion:completion];
+               completion:^{
+                   [weakSelf updateCharts];
+               }];
+    }
+    
+    //update the correlatedScoring object
+    if (self.correlatedScoring) {
+        [self.correlatedScoring updatePeriodForDays:numberOfDays groupBy:groupBy];
     }
+    
+}
+
+//should only be called by parent self of correlated data source
+- (void)correlateDataSources{
+    //move dataPoints into correlateDataPoints
+    [self discardIncongruentArrayElements];
+    
+    //index the arrays
+    [self indexDataSeries:self.dataPoints];
+    [self indexDataSeries:self.correlatedScoring.dataPoints];
+}
+
+- (void)updateCharts{
+    
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if (self.correlatedScoring) {
+            [self correlateDataSources];
+        }else if (self.weakParentScoring){
+            [self.weakParentScoring correlateDataSources];
+        }
+        
+        [[NSNotificationCenter defaultCenter]postNotificationName:APCUpdateChartsNotification object:self];
```

Comment:
Why is this a notification and not a message to a delegate with a proper protocol? Surely only charts for which this is the source need to know that these charts have changed.